functions/middleware/auth.js

```diff
@@ -0,0 +1,92 @@
+const express = require('express');
+const admin = require('firebase-admin');
+const logger = require('./logger')
+const constants = require('../common/constants')
+
+/**
+ * Middleware to verify whether the user is authenticated
+ * @param {request} request 
+ * @param {response} response 
+ * @param {function to pass control} next
+ */
+const isAuthenticated = async function (request, response, next) {
+    const errMessage = 'Authorization token not found in header or is in invalid format!!'
+    const { authorization } = request.headers
+    try {
+        if (!authorization || !authorization.startsWith('Bearer')) {
+            logger.warn(errMessage)
+            throw new Error(constants.UNAUTHORIZED)
+        }
+        const split = authorization.split('Bearer ')
+        if (split.length !== 2) {
+            logger.warn(errMessage)
+            throw new Error(constants.UNAUTHORIZED)
+        }
+        const authToken = split[1]
+        //const decodedToken = await admin.auth().verifyIdToken(authToken)
+        //logger.info(`User ${decodedToken.firstName} is authenticated`)
+        // request.user = { 
+        //     uid: decodedToken.uid, 
+        //     role: customClaims.role, 
+        //     branch: customClaims.branch, 
+        //     firstName: decodedToken.firstName, 
+        //     lastName: decodedToken.lastName,
+        //     email: decodedToken.email 
+        // }
+
+        // TODO: Remove this
+        request.user = {
+            uid: '5OGy1Jo2JXf2nXocHEVtZX7116K2',
+            role: 'admin',
+            branch: 'kormangala',
+            firstName: 'Walter',
+            lastName: 'White',
+            email: 'walter@white.com'
+        }
+
+        request.user = {
+            uid: '5OGy1Jo2JXf2nXocHEVtZX7116K2',
+            role: 'admin',
+            branch: 'kormangala',
+            firstName: 'Jesse',
+            lastName: 'Pinkman',
+            email: 'jesse@pinkman.com'
+        }
+    } catch (err) {
+        // logger.error(`In Catch of IsAuthenticated`)
+        // logger.error(`${err.message}`)
+        // err.message = constants.UNAUTHORIZED
+        // err.statusCode = 401
+        // next(err)
+        // return;
+        throw err
+    }
+};
+
+/**
+ * Method to check if user is Admin
+ * @param {request} request 
+ * @param {response} response 
+ * @param {function to pass control} next
+ */
+const isAdmin = async function (request, response, next) {
+    isAuthenticated(request, response, next).then(() => {
+        const role = request.user.role
+        logger.debug(`Current user has role ${role}`)
+        if (!role || role !== constants.ADMIN) {
+            logger.warn(`Current user is not a admin and has role ${role}`);
+            throw new Error(`Unauthorized`)
+        } else {
+            next()
+        }
+    }).catch(err => {
+        err.statusCode = 401
```

Comment:
as per my understanding 403 is for unauthorized and 401 is for unauthenticated. Please double check

Reply:
I had checked it before using it. 403 is forbidden and it is unrelated to credentials as per my understanding.

Reply:
umm, role based access , 403 is right i guess. So if you are a system user and you try to create a user , you are forbidden to perform the action.